File: apollo-portal/src/main/resources/static/scripts/controller/ServerConfigController.js
/*
 * Copyright 2022 Apollo Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
server_config_manage_module.controller('ServerConfigController',
    ['$scope', '$window', '$translate', 'toastr', 'AppUtil', 'ServerConfigService', 'PermissionService',
        'EnvService', ServerConfigController]);

function ServerConfigController($scope, $window, $translate, toastr, AppUtil, ServerConfigService, PermissionService, EnvService) {

    $scope.serverConfig = {};
    $scope.portalDBConfigs = [];
    $scope.portalDBFilterConfigs = [];
    $scope.configDBConfigs = [];
    $scope.configDBFilterConfigs = [];
    $scope.envs = [];
    $scope.selectedEnv = '';
    $scope.displayModule = 'home';
    $scope.portalDBConfigSearchKey = '';
    $scope.configDBConfigSearchKey = '';
    $scope.configEdit = configEdit;
    $scope.createPortalDBConfig = createPortalDBConfig;
    $scope.createConfigDBConfig = createConfigDBConfig;
    $scope.gobackPortalDBTabs = gobackPortalDBTabs;
    $scope.gobackConfigDBTabs = gobackConfigDBTabs;
    $scope.portalDBConfigFilter = portalDBConfigFilter;
    $scope.configDBConfigFilter = configDBConfigFilter;
    $scope.resetPortalDBConfigSearchKey = resetPortalDBConfigSearchKey;
    $scope.resetConfigDBConfigSearchKey = resetConfigDBConfigSearchKey;
    $scope.switchConfigDBEnvs = switchConfigDBEnvs;

    $scope.allowSwitchingTabs = true;

    init();

    function init() {
        initPermission();
        getPortalDBConfig();
        initEnv();
    }

    function initEnv() {
        EnvService.find_all_envs().then(function (result) {
            $scope.envs = result;
            $scope.selectedEnv = result[0];
            getConfigDBConfig();
        });
    }

    function initPermission() {
        PermissionService.has_root_permission()
            .then(function (result) {
                $scope.isRootUser = result.hasPermission;
            });
    }

    function getPortalDBConfig() {
        ServerConfigService.findPortalDBConfig()
            .then(function (result) {
                $scope.portalDBConfigs = [];
                $scope.portalDBFilterConfigs = [];
                result.forEach(function (user) {
                    $scope.portalDBConfigs.push(user);
                    $scope.portalDBFilterConfigs.push(user);
                });
            }, function (result) {
                $scope.portalDBConfigs = [];
                $scope.portalDBFilterConfigs = [];
                toastr.error(AppUtil.errorMsg(result), $translate.instant('Config.SystemError'));
            });
    }

    function getConfigDBConfig() {
        ServerConfigService.findConfigDBConfig($scope.selectedEnv)
            .then(function (result) {
                $scope.configDBConfigs = [];
                $scope.configDBFilterConfigs = [];
                result.forEach(function (user) {
                    $scope.configDBConfigs.push(user);
                    $scope.configDBFilterConfigs.push(user);
                });
            }, function (result) {
                $scope.configDBConfigs = [];
                $scope.configDBFilterConfigs = [];
                toastr.error(AppUtil.errorMsg(result), $translate.instant('Config.SystemError'));
            });
    }

    function configEdit(displayModule, config) {
        $scope.displayModule = displayModule;
        $scope.allowSwitchingTabs = false;

        $scope.serverConfig = {};
        if (config != null) {
            $scope.serverConfig = {
                key: config.key,
                value: config.value,
                comment: config.comment
            };
        }
    }

    function switchConfigDBEnvs(env) {
        $scope.selectedEnv = env;
        getConfigDBConfig();
    }

    function createPortalDBConfig() {
        ServerConfigService.createPortalDBConfig($scope.serverConfig).then(function (result) {
            toastr.success($translate.instant('ServiceConfig.Saved'));
            getPortalDBConfig();
            $scope.displayModule = 'home';
            $scope.allowSwitchingTabs = true;
        }, function (result) {
            toastr.error(AppUtil.errorMsg(result), $translate.instant('ServiceConfig.SaveFailed'));
        });
    }

    function createConfigDBConfig() {
        ServerConfigService.createConfigDBConfig($scope.selectedEnv, $scope.serverConfig).then(function (result) {
            toastr.success($translate.instant('ServiceConfig.Saved'));
            getConfigDBConfig();
            $scope.displayModule = 'home';
            $scope.allowSwitchingTabs = true;
        }, function (result) {
            toastr.error(AppUtil.errorMsg(result), $translate.instant('ServiceConfig.SaveFailed'));
        });
    }


    function gobackPortalDBTabs() {
        $scope.displayModule = 'home';
        $scope.allowSwitchingTabs = true;
        getPortalDBConfig();
    }

    function gobackConfigDBTabs() {
        $scope.displayModule = 'home';
        $scope.allowSwitchingTabs = true;
        getConfigDBConfig();
    }

    function portalDBConfigFilter() {
        $scope.portalDBConfigSearchKey = $scope.portalDBConfigSearchKey.toLowerCase();
        let filterConfig = [];
        $scope.portalDBConfigs.forEach(function (item) {
            let keyName = item.key;
            if (keyName && keyName.toLowerCase().indexOf($scope.portalDBConfigSearchKey) >= 0) {
                filterConfig.push(item);
            }
        });
        $scope.portalDBFilterConfigs = filterConfig;
    }

    function resetPortalDBConfigSearchKey() {
        $scope.portalDBConfigSearchKey = '';
        portalDBConfigFilter();
    }

    function configDBConfigFilter() {
        $scope.configDBConfigSearchKey = $scope.configDBConfigSearchKey.toLowerCase();
        let filterConfig = [];
        $scope.configDBConfigs.forEach(function (item) {
            let keyName = item.key;
            if (keyName && keyName.toLowerCase().indexOf($scope.configDBConfigSearchKey) >= 0) {
                filterConfig.push(item);
            }
        });
        $scope.configDBFilterConfigs = filterConfig;
    }

    function resetConfigDBConfigSearchKey() {
        $scope.configDBConfigSearchKey = '';
        configDBConfigFilter();
    }
}

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/ServerConfigController.java
/*
 * Copyright 2022 Apollo Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.ctrip.framework.apollo.portal.controller;


import com.ctrip.framework.apollo.portal.entity.po.ServerConfig;
import com.ctrip.framework.apollo.portal.environment.Env;
import com.ctrip.framework.apollo.portal.service.ServerConfigService;

import java.util.List;
import javax.validation.Valid;

import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

/**
 * 配置中心本身需要一些配置,这些配置放在数据库里面
 */
@RestController
public class ServerConfigController {
    private final ServerConfigService serverConfigService;

    public ServerConfigController(final ServerConfigService serverConfigService) {
        this.serverConfigService = serverConfigService;
    }

    @PreAuthorize(value = "@permissionValidator.isSuperAdmin()")
    @PostMapping("/server/portal-db/config")
    public ServerConfig createOrUpdatePortalDBConfig(@Valid @RequestBody ServerConfig serverConfig) {
        return serverConfigService.createOrUpdatePortalDBConfig(serverConfig);
    }

    @PreAuthorize(value = "@permissionValidator.isSuperAdmin()")
    @PostMapping("/server/envs/{env}/config-db/config")
    public ServerConfig createOrUpdateConfigDBConfig(@Valid @RequestBody ServerConfig serverConfig, @PathVariable String env) {
        return serverConfigService.createOrUpdateConfigDBConfig(Env.transformEnv(env), serverConfig);
    }

    @PreAuthorize(value = "@permissionValidator.isSuperAdmin()")
    @GetMapping("/server/portal-db/config/find-all-config")
    public List<ServerConfig> findAllPortalDBServerConfig() {
        return serverConfigService.findAllPortalDBConfig();
    }

    @PreAuthorize(value = "@permissionValidator.isSuperAdmin()")
    @GetMapping("/server/envs/{env}/config-db/config/find-all-config")
    public List<ServerConfig> findAllConfigDBServerConfig(@PathVariable String env) {
        return serverConfigService.findAllConfigDBConfig(Env.transformEnv(env));
    }

}
